src/snowflake/snowpark/profiler.py

```diff
@@ -0,0 +1,210 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+import re
+from contextlib import contextmanager
+from typing import List, Optional
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.utils import validate_object_name
+
+
+class Profiler:
+    """
+    Setup profiler to receive profiles of stored procedures.
+
+    Note:
+        This feature cannot be used in owner's right SP because owner's right SP will not be able to set session-level parameters.
+    """
+
+    def __init__(
+        self,
+        stage: Optional[str] = "",
+        active_profiler: Optional[str] = "LINE",
+        session: Optional["snowflake.snowpark.Session"] = None,
```

Comment:
what should customer expect to happen if session is None?

Reply:
session is required for profiler to work.
There is two way to use profiler as described in the design doc. 
In context manager, session needed to be provided when create the context manager.
In profiler class, session is set when registering profiler.
This is why I make it optional here.